File: backend/python/tests/test_solver.py
#!/usr/bin/env python3
"""
Test script for the Thunder Scheduler Constraint Solver
"""

import sys
import os
import json

# Add the parent directory to the path so we can import the solver modules
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from solver.constraint_solver import ScheduleSolver
from solver.solution_validator import ScheduleValidator


def create_test_data():
    """Create a small test dataset for the solver."""
    return {
        'classes': [
            {'id': 'class1', 'name': 'Class 1A', 'gradeLevel': 1},
            {'id': 'class2', 'name': 'Class 1B', 'gradeLevel': 1},
            {'id': 'class3', 'name': 'Class 2A', 'gradeLevel': 2},
            {'id': 'class4', 'name': 'Class 2B', 'gradeLevel': 2},
        ],
        'conflicts': {
            'class1': {
                'MONDAY': [1, 2, 3],
                'TUESDAY': [6, 7],
                'WEDNESDAY': [4, 5],
                'THURSDAY': [2, 3],
                'FRIDAY': [7, 8]
            },
            'class2': {
                'MONDAY': [4, 5, 6],
                'TUESDAY': [1, 2],
                'WEDNESDAY': [6, 7],
                'THURSDAY': [4, 5],
                'FRIDAY': [1, 2]
            },
            'class3': {
                'MONDAY': [7, 8],
                'TUESDAY': [3, 4, 5],
                'WEDNESDAY': [2, 3],
                'THURSDAY': [6, 7, 8],
                'FRIDAY': [3, 4]
            },
            'class4': {
                'MONDAY': [2, 3],
                'TUESDAY': [7, 8],
                'WEDNESDAY': [1, 2, 3],
                'THURSDAY': [4, 5],
                'FRIDAY': [6, 7]
            }
        },
        'teacherAvailability': {
            'MONDAY': [1, 8],  # Teacher unavailable during periods 1 and 8 on Monday
            'FRIDAY': [7, 8]   # Teacher unavailable during periods 7 and 8 on Friday
        },
        'constraints': {
            'maxClassesPerDay': 3,
            'maxClassesPerWeek': 12,
            'maxConsecutiveClasses': 2,
            'requireBreakAfterClass': True
        },
        'rotationWeeks': 1
    }


def test_solver():
    """Test the constraint solver with a small dataset."""
    print("Testing constraint solver...")
    
    # Create test data
    test_data = create_test_data()
    
    # Create solver
    solver = ScheduleSolver(test_data)
    
    # Solve and get result
    result = solver.solve()
    
    # Print result
    print(f"Solver status: {result['status']}")
    print(f"Solve time: {result['solveTime']:.2f} seconds")
    
    if result['status'] == 'success':
        print(f"Found {len(result['solution'])} assignments")
        
        # Print assignments in a readable format
        assignments_by_day = {}
        for assignment in result['solution']:
            day = assignment['day']
            if day not in assignments_by_day:
                assignments_by_day[day] = []
            assignments_by_day[day].append(assignment)
        
        for day in ['MONDAY', 'TUESDAY', 'WEDNESDAY', 'THURSDAY', 'FRIDAY']:
            if day in assignments_by_day:
                print(f"\n{day}:")
                day_assignments = sorted(assignments_by_day[day], key=lambda a: a['period'])
                for assignment in day_assignments:
                    class_id = assignment['classId']
                    period = assignment['period']
                    class_name = next((c['name'] for c in test_data['classes'] if c['id'] == class_id), class_id)
                    print(f"  Period {period}: {class_name}")
        
        # Validate the solution
        validator = ScheduleValidator({
            'assignments': result['solution'],
            'classes': test_data['classes'],
            'conflicts': test_data['conflicts'],
            'teacherAvailability': test_data['teacherAvailability'],
            'constraints': test_data['constraints']
        })
        validation_result = validator.validate()
        
        print(f"\nValidation result: {'Valid' if validation_result['valid'] else 'Invalid'}")
        if not validation_result['valid']:
            print(f"Violations: {len(validation_result['violations'])}")
            for violation in validation_result['violations']:
                print(f"  - {violation['message']}")
    else:
        print(f"No solution found: {result['message']}")


if __name__ == "__main__":
    test_solver()

File: backend/src/test-solver.ts
import { PrismaClient } from '@prisma/client';
import { ClassService } from './services/class.service';
import { ScheduleService } from './services/schedule.service';
import { SolverService } from './services/solver.service';
import { ScheduleGenerationRequest, ScheduleConstraints } from './types';
import fs from 'fs';
import path from 'path';

// Initialize Prisma Client
const prisma = new PrismaClient();

// Initialize services
const classService = new ClassService(prisma);
const scheduleService = new ScheduleService(prisma);
const solverService = new SolverService();

/**
 * Test the solver with a small dataset
 */
async function testSolver() {
  try {
    console.log('Starting solver test...');
    
    // Clear existing data
    console.log('Clearing existing data...');
    await prisma.assignment.deleteMany();
    await prisma.schedule.deleteMany();
    await prisma.conflict.deleteMany();
    await prisma.class.deleteMany();
    await prisma.teacherAvailability.deleteMany();
    
    // Import test data
    console.log('Importing test data...');
    const csvPath = path.join(__dirname, '../../test_data/small-test-data.csv');
    const csvData = fs.readFileSync(csvPath, 'utf8');
    await classService.importClassesFromCSV(csvData);
    
    // Get all classes
    const classes = await classService.getAllClasses();
    console.log(`Imported ${classes.length} classes`);
    
    // Create some teacher availability
    console.log('Creating teacher availability...');
    const today = new Date();
    const startDate = new Date(today);
    startDate.setDate(today.getDate() - today.getDay() + 1); // Start from Monday
    
    // Create teacher unavailability for Monday morning and Friday afternoon
    await scheduleService.createTeacherAvailability({
      date: new Date(startDate),
      blockedPeriods: [1, 2],
      reason: 'Staff meeting'
    });
    
    const fridayDate = new Date(startDate);
    fridayDate.setDate(startDate.getDate() + 4); // Friday
    await scheduleService.createTeacherAvailability({
      date: fridayDate,
      blockedPeriods: [7, 8],
      reason: 'Early dismissal'
    });
    
    // Set up schedule generation request
    const endDate = new Date(startDate);
    endDate.setDate(startDate.getDate() + 4); // 5 days (Monday to Friday)
    
    const constraints: ScheduleConstraints = {
      maxClassesPerDay: 3,
      maxClassesPerWeek: 12,
      maxConsecutiveClasses: 2,
      requireBreakAfterClass: true
    };
    
    const request: ScheduleGenerationRequest = {
      startDate,
      endDate,
      rotationWeeks: 1,
      constraints
    };
    
    // Generate schedule
    console.log('Generating schedule...');
    console.log(`Start date: ${startDate.toISOString()}`);
    console.log(`End date: ${endDate.toISOString()}`);
    console.log('Constraints:', JSON.stringify(constraints, null, 2));
    
    const schedule = await scheduleService.generateSchedule(request);
    
    // Print results
    console.log('\nSchedule generated successfully!');
    console.log(`Schedule ID: ${schedule.id}`);
    console.log(`Number of assignments: ${schedule.assignments.length}`);
    
    // Group assignments by day
    const assignmentsByDay: Record<string, any[]> = {};
    schedule.assignments.forEach(assignment => {
      if (!assignmentsByDay[assignment.day]) {
        assignmentsByDay[assignment.day] = [];
      }
      assignmentsByDay[assignment.day].push(assignment);
    });
    
    // Print schedule by day
    for (const day of ['MONDAY', 'TUESDAY', 'WEDNESDAY', 'THURSDAY', 'FRIDAY']) {
      if (assignmentsByDay[day]) {
        console.log(`\n${day}:`);
        const sortedAssignments = assignmentsByDay[day].sort((a, b) => a.period - b.period);
        
        for (const assignment of sortedAssignments) {
          const classInfo = classes.find(c => c.id === assignment.classId);
          console.log(`  Period ${assignment.period}: ${classInfo?.name || assignment.classId}`);
        }
      }
    }
    
    // Validate schedule
    console.log('\nValidating schedule...');
    const validationResult = await scheduleService.validateSchedule(schedule.id, constraints);
    
    if (validationResult.valid) {
      console.log('Schedule is valid!');
    } else {
      console.log('Schedule has violations:');
      validationResult.violations.forEach(violation => {
        console.log(`  - ${violation.message}`);
      });
    }
    
    console.log('\nTest completed successfully!');
  } catch (error) {
    console.error('Error during test:', error);
  } finally {
    // Disconnect Prisma
    await prisma.$disconnect();
  }
}

// Run the test
testSolver().catch(console.error);